vng_api_common/extensions/field_extensions.py

```diff
@@ -0,0 +1,357 @@
+import logging
+
+from django.utils.translation import gettext_lazy as _
+
+from drf_spectacular.extensions import OpenApiSerializerFieldExtension
+from drf_spectacular.openapi import AutoSchema, OpenApiTypes
+from drf_spectacular.plumbing import ResolvedComponent, build_basic_type
+
+logger = logging.getLogger(__name__)
+
+TYPES_MAP = {
+    str: OpenApiTypes.STR,
+    int: OpenApiTypes.INT,
+    bool: OpenApiTypes.BOOL,
+}
+
+
+class ReadOnlyFieldExtension(OpenApiSerializerFieldExtension):
+    target_class = "rest_framework.serializers.ReadOnlyField"
+
+    def map_serializer_field(self, auto_schema: AutoSchema, direction):
+        prop = getattr(self.target.parent.Meta.model, self.target.source)
+        if not isinstance(prop, property):
+            return super().map_serializer_field(auto_schema, direction)
+
+        return_type = prop.fget.__annotations__.get("return")
+        if return_type is None:
+            logger.debug(
+                "Missing return type annotation for prop %s on model %s",
+                self.field.source,
+                self.field.parent.Meta.model,
+            )
+            return super().map_serializer_field(auto_schema, direction)
+
+        type_ = TYPES_MAP.get(return_type)
+        if type_ is None:
+            logger.debug("Missing type mapping for %r", return_type)
+
+        return build_basic_type(type_ or OpenApiTypes.STR)
+
+
+class HyperlinkedRelatedFieldExtension(OpenApiSerializerFieldExtension):
+    target_class = "vng_api_common.serializers.LengthHyperlinkedRelatedField"
+    match_subclasses = True
+
+    def map_serializer_field(self, auto_schema: AutoSchema, direction):
+        return {
+            "type": "string",
+            "format": "uri",
+            "minLength": self.target.min_length,
+            "maxLength": self.target.max_length,
+            "description": self.target.help_text or "",
+        }
+
+
+class HyperlinkedIdentityFieldExtension(OpenApiSerializerFieldExtension):
+    target_class = "rest_framework.serializers.HyperlinkedIdentityField"
+
```

Comment:
```suggestion
match_subclasses = True
```

If there is no specific reason to exclude subclasses, I'd add them